omnipaxos/src/omni_paxos.rs

```diff
@@ -24,87 +26,185 @@ use toml;
     where
         T: Entry,
         B: Storage<T>,
     {
-        assert_ne!(self.pid, 0, "Pid cannot be 0");
-        assert_ne!(self.configuration_id, 0, "Configuration id cannot be 0");
-        assert!(!self.peers.is_empty(), "Peers cannot be empty");
-        assert!(
-            !self.peers.contains(&self.pid),
-            "Peers should not include self pid"
-        );
-        assert!(
-            self.batch_size >= 1,
-            "Batch size must be greater than or equal to 1"
-        );
-        assert!(self.buffer_size > 0, "Buffer size must be greater than 0");
-
+        self.validate()?;
         // Use stored ballot as initial BLE leader
         let recovered_leader = storage
             .get_promise()
             .expect("storage error while trying to read promise");
-        OmniPaxos {
+        Ok(OmniPaxos {
             ble: BallotLeaderElection::with(self.clone().into(), recovered_leader),
-            election_clock: LogicalClock::with(self.election_tick_timeout),
-            resend_message_clock: LogicalClock::with(self.resend_message_tick_timeout),
+            election_clock: LogicalClock::with(self.server_config.election_tick_timeout),
+            resend_message_clock: LogicalClock::with(
+                self.server_config.resend_message_tick_timeout,
+            ),
             seq_paxos: SequencePaxos::with(self.into(), storage),
+        })
+    }
+}
+
+/// Configuration for an `OmniPaxos` cluster.
+/// # Fields
+/// * `configuration_id`: The identifier for the cluster configuration that this OmniPaxos server is part of.
+/// * `nodes`: The nodes in the cluster i.e. the `pid`s of the other servers in the configuration.
+/// * `flexible_quorum` : Defines read and write quorum sizes. Can be used for different latency vs fault tolerance tradeoffs.
+#[derive(Clone, Debug, PartialEq, Default)]
+#[cfg_attr(any(feature = "serde", feature = "toml_config"), derive(Deserialize))]
+#[cfg_attr(feature = "toml_config", serde(default))]
+#[cfg_attr(feature = "serde", derive(Serialize))]
+pub struct ClusterConfig {
+    /// The identifier for the cluster configuration that this OmniPaxos server is part of. Must
+    /// not be 0 and be greater than the previous configuration's id.
+    pub configuration_id: ConfigurationId,
+    /// The nodes in the cluster i.e. the `pid`s of the other servers in the configuration.
+    pub nodes: Vec<NodeId>,
+    /// Defines read and write quorum sizes. Can be used for different latency vs fault tolerance tradeoffs.
+    pub flexible_quorum: Option<FlexibleQuorum>,
+}
+
+impl ClusterConfig {
+    /// Checks that all the fields of the cluster config are valid
+    fn validate(&self) -> Result<(), ConfigError> {
+        let num_nodes = self.nodes.len();
+        valid_config!(num_nodes > 1, "Need more than 1 node");
+        valid_config!(self.configuration_id != 0, "Configuration ID cannot be 0");
+        if let Some(FlexibleQuorum {
```

Comment:
I moved the FlexibleQuorum check back here. Before the validity of the flexible quorum was only checked when we constructed SequencePaxos, which meant a user could propose a SS with an invalid flexible quorum. The check cant be done when creating the FlexibleQuorum struct since we need to know the # of nodes.